habitat_baselines/common/obs_transformers.py

```diff
@@ -224,264 +223,108 @@ class Cube2Equirec(nn.Module):
-        orientation_mask = (
-            mask_back * 0
-            + mask_down * 1
-            + mask_front * 2
-            + mask_left * 3
-            + mask_right * 4
-            + mask_up * 5
-        )
-
-        return (
-            torch.cat(
-                [
-                    grid_back,
-                    grid_down,
-                    grid_front,
-                    grid_left,
-                    grid_right,
-                    grid_up,
-                ],
-                0,
-            ),
-            orientation_mask,
-        )
-
-    # Convert cubic images to equirectangular
-    def _to_equirec(self, batch: torch.Tensor):
+        self.grids = self.generate_grid(equ_h, equ_w)
+
+    def generate_grid(self, equ_h: int, equ_w: int) -> torch.Tensor:
+        # Project on sphere
+        theta_map, phi_map = self.get_theta_phi_map(equ_h, equ_w)
+        xyz_on_sphere = self.angle2sphere(theta_map, phi_map)
+
+        # Rotate so that each face will be in front of camera
+        rotations = [
+            np.array([[-1, 0, 0], [0, 1, 0], [0, 0, -1]]),  # Back
+            np.array([[1, 0, 0], [0, 0, 1], [0, -1, 0]]),  # Down
+            np.array([[1, 0, 0], [0, 1, 0], [0, 0, 1]]),  # Front
+            np.array([[0, 0, -1], [0, 1, 0], [1, 0, 0]]),  # Left
+            np.array([[0, 0, 1], [0, 1, 0], [-1, 0, 0]]),  # Right
+            np.array([[1, 0, 0], [0, 0, -1], [0, 1, 0]]),  # Up
+        ]
+        # Generate grid
+        grids = []
+        for rot in rotations:
+            R = torch.from_numpy(rot.T).float()
+            rotate_on_sphere = torch.matmul(
+                xyz_on_sphere.view((-1, 3)), R
+            ).view(equ_h, equ_w, 3)
+
+            # Project points on z=1 plane
+            grid = rotate_on_sphere / torch.abs(rotate_on_sphere[..., 2:3])
+            mask = torch.abs(grid).max(-1)[0] <= 1  # -1 <= grid.xy <= 1
+            mask *= grid[..., 2] == 1
+            grid[
+                ~mask
+            ] = 2  # values bigger than one will be ignored by grid_sample
+            grid_xy = -grid[..., :2].unsqueeze(0)
+            grids.append(grid_xy)
+        grids = torch.cat(grids, dim=0)
+        return grids
+
+    def _to_equirec(self, batch: torch.Tensor) -> torch.Tensor:
+        """Takes a batch of cubemaps stacked in proper order and converts thems to equirects, reduces batch size by 6"""
         batch_size, ch, _H, _W = batch.shape
-        if batch_size != 6:
-            raise ValueError("Batch size mismatch!!")
-
-        output = torch.zeros(
-            1, ch, self.equ_h, self.equ_w, device=batch.device
+        cubemap_sides = batch.size()[0]
+        if cubemap_sides == 0 or cubemap_sides % 6 != 0:
+            raise ValueError("Batch size should be 6x")
+        output = torch.nn.functional.grid_sample(
+            batch,
+            self.grids.repeat(batch_size // 6, 1, 1, 1),
+            align_corners=True,
```

Comment:
Align_corners here is nice, but it does import a performance penalty if I recall, is it worth it?